### PR TITLE
opt(server): Execute lua on target shard, if it's 1

### PR DESCRIPTION
**Pull request by chakaz**
This will save hops by short-circuiting execution of commands.

### CODE REVIEW

**Review by dranikpg — CHANGES_REQUESTED** [review]
I surely wouldn't do it this way

1. If you look ten lines above, you'll see that `StartMultiEval` already scheduled the multi transaction and computed the shards it needs to run on. You can ask it for its unique_shard_id and unique_shard_count to check if its only a single shard

2. I wouldn't switch fibers for transactions that run non-atomically or globally, you can't rely on the shard info for them (non atomic starts off empty)

3. I just noticed we don't check inline scheduling conditions for the short circuit in ExecuteAsync

Please tag me in further PRs that change transaction stuff

**Comment by romange** [comment]
@dranikpg for the use-case we check, the codepath should not go via ExecuteAsync.

**Comment by dranikpg** [comment]
> @dranikpg for the use-case we check, the codepath should not go via ExecuteAsync.

You mean the opposite - schedule single hop 🙃 It will for non atomic. I know its never used but lets not break stuff, its difficult to fix later

File: src/server/main_service.cc
Comment (romange) [inline src/server/main_service.cc:1462]: @chakaz can you please loadtest the server on 16-threads (4xlarge) and attach here the profilez graph?
also, can you say how much CPU utilization approximately it reaches at peak during the load?
I am afraid we spend lots of CPU on launching a new fiber for every `Await` call, and I would like to check it.

**Comment by dranikpg** [comment]
If you benchmark your changes, please disable squashing (you will trigger it very easily with a loop). We should probably disable squashing at all in this case because it offers no benefits - there is no parallelization opportunity, but it will try to find one. The time it saves on hops is made worse by the fact that it delayed execution

File: src/server/main_service.cc
Comment (chakaz) [reply src/server/main_service.cc:1462]: Sure thing, but won't happen today I'm afraid :sleepy:

**Comment by chakaz** [comment]
I'll disable squashing in a follow-up PR (and will benchmark with that)

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:1448]: I just remembered: if we move ove the execution to another thread, don't we access the connection context from another thread as well? We need to be safe in that case because we might have a dispatch fiber still working with it

**Comment by chakaz** [comment]
> 3. I just noticed we don't check inline scheduling conditions for the short circuit in ExecuteAsync
> 4. We also have a bug around coordinator_index_ 💀
Just making sure - you fixed both or these and no further AIs required on (3) and (4), right?

File: src/server/main_service.cc
Comment (chakaz) [reply src/server/main_service.cc:1448]: Wow, that's an excellent comment.
In what circumstances could there be other fibers active? In the case of replication, maybe?
It sounds like a hard problem to solve (pausing or waiting for other fibers because moving execution). Do you have any advice?

Comment (dranikpg) [reply src/server/main_service.cc:1448]: > In what circumstances could there be other fibers active? In the case of replication, maybe?

In case of pipelined execution, one fiber reads, the other executes, they communicate over one queue: the `dispatch_q` in code

---

It is unsafe now in some cases. For example, if the monitor is enabled, monitor messages will be added to the dispatch queue from this other thread of execution. If this EVAL command was part of a pipeline, then the dispatch queue is accessed by the reader fiber at the same time -> race

That was about EVAL only ⬆️, not the commands it executes. If you execute any commands that access the context mutably, it might break as well.

---

How does squashing work? Squashing creates context copies and captures replies to re-order them safely. Other stuff like monitoring or even mutating the context doesn't work there from non-native threads.

--- 

> Do you have any advice?

Not yet. I'm working on this (that's why I remembered 😁 ) - but for all commands to parallelize pipelines. Once I solved it for my case, it'll be solved for this as well

**Comment by dranikpg** [comment]
> Just making sure - you fixed both or these and no further AIs required on (3) and (4), right?

Yes

**Review by dranikpg — APPROVED** [review]
I can give you an approve on this because there is no simple solution to the issue discussed above. We shouldn't let this into a release, yet its unlikely someone will discover it on main

You can modify `Connection::SendAsync` to disallow dispatching from other threads. This should be the only trouble point of trouble for your case.

**Comment by chakaz** [comment]
It sounds like it will be better to wait until your solution is merged and use it, right?
I'd rather not let main have anything not planned for a release, especially not before I go OOO for a few weeks :)

**Comment by romange** [comment]
@chakaz  I think it's still better to benchmark it and check for the cpu bottlenecks in parallel.

**Comment by romange** [comment]
@dranikpg what's the specific scenario that can break this PR?

**Comment by dranikpg** [comment]
@dranikpg what's the specific scenario that can break this PR?

For example enable monitor and then execute this script as part of a pipeline with more commands behind it

**Comment by dranikpg** [comment]
We can check for monitor monitor and merge this really (what if someone enables while executing 😆 🤔 ). In a few days I'll have checked all context accesses so I can fix this if needed then

**Comment by chakaz** [comment]
Not blocking on benchmarking is a good idea, which I should have thought of myself :|

Here are the results:

Benchmarks (all done on ci6n.4xlarge Dragonfly server, with ci6n.16xlarge as client), all with only addJobs. # writers and # queues are **client configuration**, i.e. not Dragonfly's

**With run-on-remote-thread:**
32 writer threads, 32 queues: 67,291.07, 66,470.27
64 writer threads, 64 queues: 100,182.30, 96,367.67
128 writer threads, 64 queues: 85,524.57, 83,713.03

CPU profiling example: https://github.com/dragonflydb/dragonfly/assets/7992598/f6b6fbd1-e62f-4388-be7e-6cbc37d1b34d

**Without run-on-remote-thread:**
32 writer threads, 32 queues: 43,837.30, 44,094.27
64 writer threads, 64 queues: 44,773.23, 42,600.43
128 writer threads, 64 queues: 43,592.60, 44,357.33

To clarify, this is an opt-built (on my machine), with this PR patched in, and the only difference is that the _without_ version runs with a force `false` on the new if

**Comment by chakaz** [comment]
The benchmark was done **without** disabling squashing (it's not trivial to disable it for Lua), so think of these as kind of a lower bound.

**Comment by romange** [comment]
@chakaz  hmm, the profile graph unites fiber creations under the same node, so it's not clear to me how much overhead adds the Await call.

image

I suggest to do the following (it will be easier that trying to figure out the exact overhead of the Await()):
in an experimental hack-branch): 
Similarly to how `EngineShard::fiber_q_` and ` std::vector<FiberQueue*> EngineShardSet::shard_queue_`  are defined and initialized, introduce hakishly `EngineShard::fiber_q2` and  `EngineShardSet::shard_queue2`. instead of running pool Await, 
switch to `engine_shard_set->Await2()`. By doing this quick and dirtly it should take not more than 30min work.  This way you will be able to Dispatch `RunFunction` remotely on permanent queue fiber without deadlocking the tx-queue.

**Comment by chakaz** [comment]
After hours of debugging these 30m work :angel: .. I came to understand that the solution isn't quite as simple as I had originally thought.
The reason is that if there's a queue for handling remote Lua runs, and that queue is handled via a single fiber, a deadlock occurs.
Think of the case where:
* Thread1 inserts itself to the transaction queue for key X
* Thread2 also inserts itself to the transaction queue for key X (same key)
* Thread2 is faster for some reason to insert itself to the queue2 thingy that I created
* Thread1 being the slow guy adds itself to queue2
This guarantees that nothing will proceed with key X :(

Perhaps we could use some sort of pool, that's less limited in parallelism, but that's a job for tomorrow's Shahar :zzz:

**Comment by romange** [comment]
oh wow 😮 , that's too bad. it seems that the fiber model becomes limiting in this case.
In a way, what we really want to do is to run a pure asynchornous code with continuation callbacks (like with C) to jump from state to state. Then, even if you call ExecuteAsync prematurely, the framework would still call the callback in correct order only after the transaction is executed. But since our code is written synchronously we can not do it and cause deadlocks.

**Comment by romange** [comment]
Good job today! Your idea of a dynamic pool on the consumer side is the most cost effective.

**Comment by chakaz** [comment]
@adiholden, @dranikpg and I discussed this just now and decided to go ahead with the merge as we all agree that it's safe.
Further optimizations can definitely help here, and I'll try to run one of Roman's proposals before I go awol

**Pull request merged**
